module3/lessons/Week4/UpdatingResources.md

```diff
@@ -0,0 +1,219 @@
+
+HTML has been used since the early 1990s.  At that time, the internet was a much simpler place.  The form tags we still use are designed to work with only two Http methods: Get and Post.
+
+ReST was defined in the early 2000's as a way to standardized http requests.  As the internet became more complex, we started to see requests being made by applications and not just by users.  Applications could more easily create ReSTful routes and requests without using a user interface (we will learn more about this in Mod4!).  
+
+<section class='instructor-notes' markdown='1'>
+Walk through the implementation together - this would be a good opportunity to let students help drive the implementation!
+</section>
+
+
+<section class='call-to-action' markdown='1'>
+
+**With a Partner**  
+Add a test to our MoviesControllerTests that satisfies this user-story
+```
+As a User
+When I visit '/movies/1/edit'
+When I update the Title and/or Genre
+    And click 'Update Movie'
+Then I am redirected to the Movie Show Page
+    And I see the updated movie information
+```
+
+</section>
+
+
+<section class='answer' markdown='1'>
+
+### One Solution
+```c#
+[Fact]
+public async Task Update_SavesChangesToMovie()
+{
+    // Arrange
+    var context = GetDbContext();
+    var client = _factory.CreateClient();
+
+    Movie movie = new Movie { Title = "Goofy", Genre = "Comedy" };
+    context.Movies.Add(movie);
+    context.SaveChanges();
+
+    var formData = new Dictionary<string, string>
+    {
+        { "Title", "Goofy" },
+        { "Genre", "Documentary" }
+    };
+
+    // Act
+    var response = await client.PostAsync($"/movies/{movie.Id}", new FormUrlEncodedContent(formData));
+    var html = await response.Content.ReadAsStringAsync();
+
+    // Assert
+    response.EnsureSuccessStatusCode();
+    Assert.Contains("Goofy", html);
+    Assert.Contains("Documentary", html);
+    Assert.DoesNotContain("Comedy", html);
+}
+```
+
+</section>
+
+### the Order of Our Tests
+
+Before we implement our controller action, we need to address one aspect of testing - the order and timing of each of our tests.  By default, xUnit will run our tests randomly, and as quickly as possible (some tests even run at the same time!).  Generally, this is a good thing - we want our tests to be individually robust, and not rely on any _other_ test to run successfully. In this case, though, we need to control our test runs a bit more.
+
+If left to their default settings, we could have two tests running at the same time (concurrently) that would put conflicting data into our test database.  xUnit provides a simple way to group tests that need to be run strictly one at a time.  We will use an attribute to group all of our Controller tests into one collection:
+
+```c#
+// MoviesControllerTests.cs
+
+namespace MvcMovie.FeatureTests
+{
+    [Collection("Controller Tests")]
```

Comment:
Nice! Looks like you found this as a solution to the errors you posted about last week. It is interesting that it's the same mock DB we are always adding to, I wonder if we need to do any wiping clean of the DB after each test.

Reply:
Hmmm... this is an interesting distinction.  We are deleting any existing database when we `GetDbContext()` during the test... so we are kind of doing the teardown from the previous test during the arrangement of our current test 🧐 

I played around with creating a different mock database for _each_ test... and I think that could work, but it took kind of a lot of re-configuring... so I went with this approach.  I think it's something we should definitely keep an eye on and aim to improve this process as we learn more about c#/.net